### PR TITLE
[optim] support of dynamic coefficients and normalization in Muon

### PR DESCRIPTION
**Pull request by thib-s**
This pull request improves the Muon optimizer's Newton–Schulz orthogonalization step. The main improvements are the introduction of configurable coefficient presets. (including "jordan" and "polar_express"), support for multiple normalization strategies, and comprehensive tests for these new options. These changes make the optimizer more adaptable to recent research and user needs. This change closes #179250 .

## Enhancements to Newton–Schulz Orthogonalization:

* Added support for coefficient presets ("jordan" and "polar_express") and per-step custom coefficients, allowing users to select from well-known or custom strategies for the Newton–Schulz iteration.
* Introduced three normalization methods ("fro", "schatten", "aol") for input matrix normalization prior to orthogonalization, which also improves convergence.

### API and Documentation:

* Updated the Muon optimizer's API to accept coefficient presets as strings, tuples, or per-step tuples. `_zeropower_via_newtonschulz` only accepts tuple of tuple to be more compliant with `torch.compile`, so the standardization of the argument is done in the init of the class.
* Extended and clarified documentation to describe the new coefficient and normalization options, with references to recent research papers.

### Testing:

* Added a test suite for the Newton–Schulz orthogonalization function, covering coefficient presets, normalization types, matrix shapes, iteration counts, error handling, and output properties (`test/optim/test_muon.py`).

## Design choices:

There seems to be an emerging consensus toward the coefficient sets found independently by [Polar-Express](https://arxiv.org/pdf/2505.16932) and [Accelerating Newton-Schulz Iteration for Orthogonalization via Chebyshev-type Polynomials](https://arxiv.org/pdf/2506.10935). I then followed what was done in [the polar express repository](https://github.com/NoahAmsel/PolarExpress/blob/main/polar_express.py#74). In order to avoid extra computation, optimal coefficients are pre-computed and stored in `PE_COEFFICIENTS`. As noted by the authors, the coefficients become constant past 10 iterations, which enables support for an arbitrary number of iterations.

Also, this line of work points out that a better normalization of the initial matrix improves convergence. As pointed out, the use of the Schatten norm allows a faster convergence without adding cost: the Gram matrix is cached and reused for the first iteration, as done in [Turbo-Muon: Accelerating Orthogonality-Based Optimization with Pre-Conditioning](https://arxiv.org/pdf/2512.04632), which also introduces another normalization named "AOL".

In order to keep the support of the original behavior, its coefficients are stored in `JORDAN_COEFFICIENTS`.

### About defaults:

Following [Beyond the Ideal: Analyzing the Inexact Muon Update](https://arxiv.org/pdf/2510.19933), a better orthonormalization leads to a stronger performance of Muon. Under this light, the use of `PE_COEFFICIENTS` and `schatten` should yield the stronger performances (I did not set `aol` as default, as it could introduce an optimization bias in some contexts)

Importantly, the paper also states that a better orthogonalization slightly changes the optimal hyperparameters. I still think that changing the default coefficients is beneficial; according to fig 1b of the paper, the performance improvement of a better orthonormalization outweighs the change of the optimal hyperparameters. Also, this choice would further reduce the sensitivity of the optimizer with respect to learning rate, momentum, and batch size. I think users who rely on these default values will experience improved performance in their training.

### CODE REVIEW

**Comment by thib-s** [comment]
I've rebased the code to be up to date.
I also updated my comment about default parameters. I would be interested in PyTorch's guidance on this.

**Comment by thib-s** [comment]
gentle reminder for @janeyx99 or @albanD to review my code. I would be happy to have positive/negative feedback about this design :-)

**Comment by janeyx99** [comment]
@thib-s thank you for the PR and sorry for the delay in review! i've been working on stable ABI stuff and so i have this PR (among other muon ones) in my backlog to review that I plan on addressing in the next few weeks!
